hibernate-core/src/main/java/org/hibernate/type/AbstractStandardBasicType.java

```diff
@@ -364,10 +360,20 @@ public Object replace(
 			Map copyCache,
 			ForeignKeyDirection foreignKeyDirection) {
 		return ForeignKeyDirection.FROM_PARENT == foreignKeyDirection
-				? getReplacement( (T) original, (T) target, session )
+				? doReplacement( original, target, session )
 				: target;
 	}
 
+	protected Object doReplacement(Object original, Object target, SharedSessionContractImplementor session) {
+		if ( original == LazyPropertyInitializer.UNFETCHED_PROPERTY ) {
```

Comment:
Is it absolutely necessary to do these checks in `AbstractStandardBasicType`? I think what Guillaume was trying to do in his PR (on my suggestion) was to do these checks in callers of `replace`. The main reason was that:

* `BasicType` is API if you refer only to the package it's in, so users could theoretically implement it
* `LazyPropertyInitializer.UNFETCHED_PROPERTY` is SPI
* Why would an an API  user have to care about a constant lost in the middle of SPIs?

Even for SPI implementors, it's asking a lot to expect them to be aware of such "magic values" without documenting it explicitly on the method's javadoc.

Maybe the right thing to do would be to move `LazyPropertyInitializer.UNFETCHED_PROPERTY` to API and document on the javadoc of `replace` the possibility of being passed a "magic value".

Reply:
It is possible to do the check in `TypeHelper`, but it would be a bit messy, because `target` should not be replaced by `null` for Lobs that use `LEGACY_LOB_MERGE_STRATEGY`.

@yrodiere, that's a good point about users potentially implementing `doReplacement`. I'll make a quick change so it's not possible.

Reply:
That's a good idea, but my point was more that users could even implement the interface themselves (from the javadoc, it seems we don't particularly recommend using this abstract class as a base). If they do, they won't have a clue that this magic value may be passed to `replace`, and even if they figure that out, accessing the constant would mean relying on SPIs...
So I think at the very least the constant should be moved to API and mentioned in the javadoc of `replace`; ideally the type should not have to deal with this at all, but from what you are saying this would be complex. I suppose we'd have to also expose something like `supportsMerge` on `Type`?

Reply:
@yrodiere, ah, I see what you mean now; I didn't realize `LazyPropertyInitializer.UNFETCHED_PROPERTY` was an SPI.

Reply:
@yrodiere, I was trying to avoid exposing something like, `Type#supportsMerge`, especially since `LEGACY_LOB_MERGE_STRATEGY` will likely get deprecated. Honestly, I don't know if anyone even uses it. As mentioned above, it is not tested.

Reply:
I also see that in 4.2, references to `LazyPropertyInitializer.UNFETCHED_PROPERTY` (in a different package) are not present in any `Type` implementations. Instead, it's used in `TypeHelper`.

Reply:
Maybe let's not take design decisions based on something legacy?

I would be fine with having the code in `TypeHelper` and `LEGACY_LOB_MERGE_STRATEGY` not working better than it was.

Reply:
I'll create a new PR based on @gsmet's  https://github.com/hibernate/hibernate-orm/pull/2390 (which moves references to `LazyPropertyInitializer.UNFETCHED_PROPERTY` to `TypeHelper`), with changes to treat `LEGACY_LOB_MERGE_STRATEGY` properly.

Reply:
@gsmet, missed your message before I posted mine.

I'll check to see if it was working in 4.2.

Reply:
BTW, `UNFETCHED_PROPERTY` is already partially handled in `TypeHelper`. That's one of the reasons why I would like the code to be there: it's better to centralize this handling rather than spreading it everywhere.

Reply:
@gsmet, yes I agree now. 

BTW, this is kind of complicated, but I realized that when merging a transient entity with an initialized Lob that is mapped as lazy with a Lob cannot be persisted using `LEGACY_LOB_MERGE_STRATEGY` cannot possibly work. When `target` gets instantiated, the field with be uninitialized. Even if we somehow fixed `LEGACY_LOB_MERGE_STRATEGY` to return `target`, then I think Hibernate would try to persist an entity with an uninitialized value. I'm not sure what would happen in this case. I guess Hibernate would throw an exception or the property would be persisted as `null`.

I'll go back to your PR and add a comment about this.